server/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -468,11 +515,259 @@ public ScriptLanguagesInfo getScriptLanguages() {
     }
 
     public ScriptStats stats() {
-        return compiler.stats();
+        return cache.get().stats();
     }
 
     @Override
     public void applyClusterState(ClusterChangedEvent event) {
         clusterState = event.state();
     }
+
+    /**
+     * CacheUpdater updates the script cache based on dynamic settings updates.  It keeps track of which context and general
+     * settings are set to determine how to fall back on settings.
+     *
+     * If the context specific {size, expiration} are set then they are used.  Otherwise, if the general {size, expiration} are
+     * set to the non-default, then we fall back to those.  Finally, if neither are set we use the context default.
+     *
+     * If the general compilation rate is set to the special value {@code ScriptService.USE_CONTEXT_RATE_KEY}, {@code "use-context"},
+     * then context caches are used.  If the general compilation rate is set to a valid value, eg 50/1m, then the general
+     * cache is used.
+     */
+    static class CacheUpdater {
+        private final Object lock = new Object();
+        private static final Tuple<Integer, TimeValue> ZERO = new Tuple<>(0, TimeValue.ZERO);
```

Comment:
`ZERO` is pretty vague as a name here, but I see it matches TimeValue. Maybe this could be `SCRIPT_COMPILATION_RATE_ZERO` or something like that to match `SCRIPT_COMPILATION_RATE`.

Reply:
Changed.